File: disciple/src/main/java/com/monster/disciple/transformer/AbsFlowableErrorCheckTransformer.java
package com.monster.disciple.transformer;

import android.content.Context;

import com.monster.disciple.response.BaseResponse;
import com.monster.disciple.util.Util;

import org.reactivestreams.Publisher;

import java.lang.ref.WeakReference;

import io.reactivex.Flowable;
import io.reactivex.FlowableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;

public abstract class AbsFlowableErrorCheckTransformer<T extends Response<R>, R extends BaseResponse>
        implements FlowableTransformer<T, R>, IErrorCheckTransformer {

    protected WeakReference<Context> mContextRef;

    public AbsFlowableErrorCheckTransformer(final Context context) {
        mContextRef = new WeakReference<>(context);
    }

    @Override
    public Publisher<R> apply(Flowable<T> upstream) {
        return upstream
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(Util.<T, R>newInstanceFunction(this));
    }
}

File: disciple/src/main/java/com/monster/disciple/transformer/AbsFlowableErrorCheckTransformer.kt
package com.monster.disciple.transformer

import android.content.Context
import android.text.TextUtils
import com.monster.disciple.CacheBuilder
import com.monster.disciple.response.BaseResponse
import com.monster.disciple.util.getCacheFunction
import com.monster.disciple.util.newInstanceFunction
import com.monster.disciple.util.saveCacheFunction
import io.reactivex.Flowable
import io.reactivex.FlowableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.reactivestreams.Publisher
import retrofit2.Response
import java.lang.ref.WeakReference

abstract class AbsFlowableErrorCheckTransformer<T : Response<R>, R : BaseResponse<*>>(context: Context?) : FlowableTransformer<T, R>, IErrorCheckTransformer {

    private var mContextRef: WeakReference<Context>? = WeakReference<Context>(context)

    private var mCacheBuilder: CacheBuilder? = null

    constructor(context: Context?, cacheBuilder: CacheBuilder?) : this(context) {
        mCacheBuilder = cacheBuilder
    }

    override fun apply(upstream: Flowable<T>): Publisher<R> {
        return if (mCacheBuilder == null || TextUtils.isEmpty(mCacheBuilder?.cacheKey)) {
            upstream
                    .map(newInstanceFunction<T, R>(this))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        } else {
            upstream
                    .map(newInstanceFunction<T, R>(this))
                    .map(saveCacheFunction<R>(mContextRef?.get(), mCacheBuilder?.cacheKey))
                    .startWith(
                            Flowable
                                    .just<String>(mCacheBuilder?.cacheKey)
                                    .map(getCacheFunction<R>(mContextRef?.get(), mCacheBuilder?.type))
                                    .filter { rCacheEntity -> !rCacheEntity.isNil() }
                                    .map { rCacheEntity -> rCacheEntity.response }
                    )
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }
}